blender rendering guide (#655)
nitrictech/docs: docs/guides/python/blender-render.mdx

Reviewer: If we are recommending uv now, this guide needs to be converted

Author: It is already using `uv`

---
description: Use the Nitric framework to build a service for rendering Blender scenes using cloud GPUs
tags:
  - API
  - AI & Machine Learning
languages:
  - python
---

# Use Cloud GPUs for rendering your Blender projects

> This example shows how you can create a remote Blender rendering application using Blender's Python interface.

Reviewer: ```suggestion
This example shows how you can create a remote [Blender](https://www.blender.org/) rendering application using Blender's Python interface.
```

We'll start by creating a new Nitric project.

<Note>
  If you want to take a look at the finished code, it can be found
  [here](https://github.com/nitrictech/examples/tree/main/v1/blender-render).
</Note>

```bash
nitric new blender-rendering py-starter
```

We can then resolve our dependencies.

```bash
uv sync
```

We'll also add [bpy](https://docs.blender.org/api/current/info_overview.html#python-in-blender) as an optional dependency. By making it an optional dependency, we can choose to only install it for the containers that require it, reducing the size of the containers. `bpy` is the API that will interact with the Blender environment for setting up and rendering our scenes. The version of `bpy` should match the version of Blender you intend on using.

```bash
uv add bpy==4.2.0 --optional ml
```

We'll organize our project structure like so:

```text
+--common/
|  +-- __init__.py
|  +-- resources.py
+--batches/
|  +-- renderer.py
+--services/
|  +-- api.py
+-- blender.dockerfile
+-- blender.dockerignore
+--.gitignore
+--.python-version
+-- pyproject.toml
+-- python.dockerfile
+-- python.dockerignore
+-- nitric.yaml
+-- README.md
```

## Creating the resources

> We'll start by creating a file to define our Nitric resources. For this project we'll need an API, Batch Job, and two buckets, one for the `.blend` files and one for the resulting renders. The API will interface with the buckets, while the Batch Job will handle the long-running renders.

Reviewer: ```suggestion
We'll start by creating a file to define our Nitric resources. For this project, we'll need an API, a batch job, and two buckets: one for the .blend files and another for the resulting renders. The API will interact with the buckets, while the batch job will handle the long-running render tasks.
```

> 

Reviewer: ```suggestion

```bash
mkdir common
touch common/resources.py
touch common/__init__.py
```

```

Reviewer: Not sure if this cut off the end

Author: I wasn't going to show the creating of each file since its shown in the project structure

  ctx.res.headers["Location"] = download_url
  ctx.res.status = 303

  return ctx

Nitric.run()
```

The final route will be for adding a `.blend` file for rendering as well as the render settings for the scene. This will add the contents of the request to a file in the blend bucket by redirecting the request to the upload URL after adding the metadata to the bucket.

```python title:services/api.py
# !collapse(1:22) collapsed
import json

from nitric.application import Nitric
from nitric.context import HttpContext
from nitric.resources import BucketNotificationContext

from common.resources import rendered_bucket, main_api, blend_bucket, renderer_job

readable_rendered_bucket = rendered_bucket.allow("read")
readable_writeable_blend_bucket = blend_bucket.allow("write", "read")
submittable_renderer_job = renderer_job.allow("submit")

@main_api.get("/render/:file")
async def get_render(ctx: HttpContext):
  file_name = ctx.req.params['file']

  download_url = await readable_writeable_blend_bucket.file(file_name).download_url(3600)

  ctx.res.headers["Location"] = download_url
  ctx.res.status = 303

  return ctx

>   @main_api.put("/:blend")

Reviewer: remove indentation, makes the code unreachable

      "fps": int(ctx.req.query.get('fps', [0])[0]),
      "device": str(ctx.req.query.get('device', ['GPU'])[0]),
      "engine": str(ctx.req.query.get('engine', ['CYCLES'])[0]),
      "animate": bool(ctx.req.query.get('animate', [False])[0]),
    }
    metadata = bytes(json.dumps(raw_metadata), encoding="utf-8")

    await readable_writeable_blend_bucket.file(f"metadata-{blend_scene_key}.json").write(metadata)

    # Write the blend scene to the bucket using an upload URL
    blend_upload_url = await readable_writeable_blend_bucket.file(f"blend-{blend_scene_key}.blend").upload_url()

    ctx.res.headers["Location"] = blend_upload_url
    ctx.res.status = 307

    return ctx

  Nitric.run()
```

We will add a storage listener which will be triggered by files being added to the `blend_bucket`. This is so we can trigger the rendering job when the rendering metadata and the `.blend` file are added to the bucket. By making this start from the listener instead of the API, we can set up workflows where rendering could be triggered from adding files to buckets manually.

```python
# !collapse(1:46) collapsed
import json

>   from nitric.application import Nitric

Reviewer: indentation

    ctx.res.headers["Location"] = download_url
    ctx.res.status = 303

    return ctx

  @main_api.put("/:blend")
  async def write_render(ctx: HttpContext):
    blend_scene_key = ctx.req.params["blend"]

    # Write the blend scene rendering settings
    raw_metadata = {
      "file_format": str(ctx.req.query.get('file_format', ['PNG'])[0]),
      "fps": int(ctx.req.query.get('fps', [0])[0]),
      "device": str(ctx.req.query.get('device', ['GPU'])[0]),
      "engine": str(ctx.req.query.get('engine', ['CYCLES'])[0]),
      "animate": bool(ctx.req.query.get('animate', [False])[0]),
    }
    metadata = bytes(json.dumps(raw_metadata), encoding="utf-8")

    await readable_writeable_blend_bucket.file(f"metadata-{blend_scene_key}.json").write(metadata)

    # Write the blend scene to the bucket using an upload URL
    blend_upload_url = await readable_writeable_blend_bucket.file(f"blend-{blend_scene_key}.blend").upload_url()

    ctx.res.headers["Location"] = blend_upload_url
    ctx.res.status = 307

    return ctx

@blend_bucket.on("write", "blend-")
async def on_written_image(ctx: BucketNotificationContext):

>   key_without_extension = ctx.req.key.split(".")[0][6:]

Reviewer: ```suggestion
  key_without_extension = os.path.splitext(ctx.req.key)[0]
```

Reviewer: This was also slicing off part of the key name itself.

Can you add the import for os to the rest of guide @HomelessDinosaur

  await submittable_renderer_job.submit(
    {
      "key": key_without_extension,
    }
  )

  return ctx

Nitric.run()
```

## Using bpy for scripted blender rendering

Start by adding our imports and the resources we defined earlier.

> 

Reviewer: ```suggestion

```bash
mkdir batches
touch batches/renderer.py
```

```

Author: I wasn't going to show the creating of each file since its shown in the project structure

  blender_bin = "blender"

  if os.path.isfile(blender_bin):
    bpy.app.binary_path = blender_bin
  else:
    ctx.res.success = False
    return ctx

  return ctx

Nitric.run()
```

Next, we'll read the blend file from the bucket that matches the key sent in the context and write it to a file accessible by the blender renderer. The line `bpy.ops.wm.open_mainfile(filepath="input")` will set the input scene for the renderer.

```python title:batches/renderer.py
# !collapse(1:10) collapsed
import os.path
import json
import glob

from nitric.context import JobContext
from nitric.application import Nitric
from common.resources import rendered_bucket, renderer_job, blend_bucket

readable_blend_bucket = blend_bucket.allow("read")
writeable_rendered_bucket = rendered_bucket.allow("write")

@renderer_job(cpus=1, memory=1024, gpus=0)
# !collapse(1:14) collapsed
async def render_image(ctx: JobContext):
  import bpy

  blend_key = ctx.req.data["key"]

  # Register the blender binary
  blender_bin = "blender"

  if os.path.isfile(blender_bin):
    bpy.app.binary_path = blender_bin
  else:
    ctx.res.success = False
    return ctx

  # load the file from a bucket to a local file

>   blend_file = await readable_blend_bucket.file(f"{blend_key}.blend").read()

Reviewer: ```suggestion
  blend_file = await readable_blend_bucket.file(f"blend-{blend_key}.blend").read()
```

  with open("input", "wb") as f:
    f.write(blend_file)

  bpy.ops.wm.open_mainfile(filepath="input")

  return ctx

Nitric.run()
```

We'll then set the settings for the render engine based on the metadata that was added to the bucket.

```python title:batches/renderer.py
# !collapse(1:10) collapsed
import os.path
import json
import glob

from nitric.context import JobContext
from nitric.application import Nitric
from common.resources import rendered_bucket, renderer_job, blend_bucket

readable_blend_bucket = blend_bucket.allow("read")
writeable_rendered_bucket = rendered_bucket.allow("write")

@renderer_job(cpus=1, memory=1024, gpus=0)
# !collapse(1:22) collapsed
async def render_image(ctx: JobContext):
  import bpy

  blend_key = ctx.req.data["key"]

  # Register the blender binary
  blender_bin = "blender"

  if os.path.isfile(blender_bin):
    bpy.app.binary_path = blender_bin
  else:
    ctx.res.success = False
    return ctx

  # load the file from a bucket to a local file

>   blend_file = await readable_blend_bucket.file(f"{blend_key}.blend").read()

Reviewer: ```suggestion
  blend_file = await readable_blend_bucket.file(f"blend-{blend_key}.blend").read()
```

  bpy.context.scene.render.filepath = blend_key
  bpy.context.scene.render.engine = metadata.get('engine')
  bpy.context.scene.cycles.device = metadata.get('device')
  bpy.context.scene.render.image_settings.file_format = metadata.get('file_format')
  bpy.context.scene.render.fps = metadata.get('fps')

  return ctx

Nitric.run()
```

The next step, is rendering depending on whether the requested scene needs to be animated or not.

```python title:batches/renderer.py
# !collapse(1:10) collapsed
import os.path
import json
import glob

from nitric.context import JobContext
from nitric.application import Nitric
from common.resources import rendered_bucket, renderer_job, blend_bucket

readable_blend_bucket = blend_bucket.allow("read")
writeable_rendered_bucket = rendered_bucket.allow("write")

@renderer_job(cpus=1, memory=1024, gpus=0)
# !collapse(1:32) collapsed
async def render_image(ctx: JobContext):
  import bpy

  blend_key = ctx.req.data["key"]

  # Register the blender binary
  blender_bin = "blender"

  if os.path.isfile(blender_bin):
    bpy.app.binary_path = blender_bin
  else:
    ctx.res.success = False
    return ctx

  # load the file from a bucket to a local file

>   blend_file = await readable_blend_bucket.file(f"{blend_key}.blend").read()

Reviewer: ```suggestion
  blend_file = await readable_blend_bucket.file(f"blend-{blend_key}.blend").read()
```

  bpy.context.scene.render.fps = metadata.get('fps')

  if metadata.get('animate'):
    bpy.ops.render.render(animation=True)
  else:
    bpy.ops.render.render(write_still=True)

  return ctx

Nitric.run()
```

With the rendering complete, we'll read the contents of the outputted render and add it to the bucket. We use a glob pattern to find the outputted file using the blend file key as the prefix.

```python title:batches/renderer.py
# !collapse(1:10) collapsed
import os.path
import json
import glob

from nitric.context import JobContext
from nitric.application import Nitric
from common.resources import rendered_bucket, renderer_job, blend_bucket

readable_blend_bucket = blend_bucket.allow("read")
writeable_rendered_bucket = rendered_bucket.allow("write")

@renderer_job(cpus=1, memory=1024, gpus=0)
# !collapse(1:37) collapsed
async def render_image(ctx: JobContext):
  import bpy

  blend_key = ctx.req.data["key"]

  # Register the blender binary
  blender_bin = "blender"

  if os.path.isfile(blender_bin):
    bpy.app.binary_path = blender_bin
  else:
    ctx.res.success = False
    return ctx

  # load the file from a bucket to a local file

>   blend_file = await readable_blend_bucket.file(f"{blend_key}.blend").read()

Reviewer: ```suggestion
  blend_file = await readable_blend_bucket.file(f"blend-{blend_key}.blend").read()
```

COPY --from=builder --chown=app:app /app /app

# Place executables in the environment at the front of the path
ENV PATH="/app/.venv/bin:$PATH"

# Run the service using the path to the handler
ENTRYPOINT python -u $HANDLER
```

Add two dockerignore files to help optimize the size of the image. These will be `python.dockerignore` and `blender.dockerignore`.

```text
.mypy_cache/
.nitric/
.venv/
nitric-spec.json
nitric.yaml
README.md
```

We can update the `nitric.yaml` file to allow our services and batch jobs to use the custom docker runtime we set up and point to the services directly.

```yaml title:nitric.yaml
name: blender-render
services:
  - match: services/*.py
    start: uv run watchmedo auto-restart -p *.py --no-restart-on-command-exit -R python -- -u $SERVICE_PATH
    runtime: python

batch-services:
  - match: batches/*.py
    start: uv run watchmedo auto-restart -p *.py --no-restart-on-command-exit -R python -- -u $SERVICE_PATH
    runtime: blender

runtimes:
  blender:
    dockerfile: blender.dockerfile
  python:
    dockerfile: python.dockerfile
```

We'll also need to add `batch-services` as a preview feature.

```yaml title:nitric.yaml
# !collapse(1:17) collapsed

> name: blender-render

Reviewer: indentation

Reviewer: would a diff + be useful for these?

Author: I think cause its adding to and not changing the lines, the default code block is fine (and they can actually copy from it).

  - python
---

# Use Cloud GPUs for rendering your Blender projects

This example shows how you can create a remote Blender rendering application using Blender's Python interface.

By using the cloud you can render your Blender scenes on infrastructure that scales and with CPU or GPU resources you might not have access to locally.

Here's a final render that was done using cloud GPUs:

<div>
  <div className="video-container">
    <iframe
      src="https://www.youtube-nocookie.com/embed/JoiAtMZ3HME"
      title="Nitric Logo Blender"
      frameBorder="0"
      allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share"
      allowFullScreen
    />
  </div>
</div>

## Prerequisites

- [UV](https://docs.astral.sh/uv/) - for dependency management
- The [Nitric CLI](/get-started/installation)
- An [AWS](https://aws.amazon.com) or [Google Cloud](https://cloud.google.com) account (_your choice_)

## Getting started

We'll start by creating a new Nitric project.

<Note>
  If you want to take a look at the finished code, it can be found
  [here](https://github.com/nitrictech/examples/tree/main/v1/blender-render).
</Note>

```bash

> nitric new blender-rendering

Reviewer: ```suggestion
nitric new blender-rendering py-starter
```

```

We can then resolve our dependencies.

```bash
uv sync
```

We'll also add [bpy](https://docs.blender.org/api/current/info_overview.html#python-in-blender) as an optional dependency. By making it an optional dependency, we can choose to only install it for the containers that require it, reducing the size of the containers. `bpy` is the API that will interact with the Blender environment for setting up and rendering our scenes. The version of `bpy` should match the version of Blender you intend on using.

```bash
uv add bpy==4.2.0 --optional ml
```

We'll organize our project structure like so:

```text
+--src/
|  +--services/

> |      +-- main.py

Reviewer 2: Was there a reason to use a `src` directory here?

readable_rendered_bucket = rendered_bucket.allow("read")
readable_writeable_blend_bucket = blend_bucket.allow("write", "read")
submittable_renderer_job = renderer_job.allow("submit")

Nitric.run()
```

We'll then write a route for getting a file from the rendered bucket. These will get a signed download url and redirect the user to this url for downloading the content.

```python title:src/services/main.py
# !collapse(1:12) collapsed
import json
import requests

from nitric.application import Nitric
from nitric.context import HttpContext
from nitric.resources import BucketNotificationContext

from src.resources import rendered_bucket, main_api, blend_bucket, renderer_job

readable_rendered_bucket = rendered_bucket.allow("read")
readable_writeable_blend_bucket = blend_bucket.allow("write", "read")
submittable_renderer_job = renderer_job.allow("submit")

@main_api.get("/render/:file")
async def get_image(ctx: HttpContext):
  file_name = ctx.req.params['file']

  download_url = await readable_writeable_blend_bucket.file(file_name).download_url(3600)

  ctx.res.headers["Location"] = download_url
  ctx.res.status = 303

  return ctx

Nitric.run()
```

> The final route will be for adding a `.blend` file for rendering as well as the render settings for the scene. This will add the contents of the request to a file in the blend bucket. Writing directly to a bucket is limited to 4 MB, to avoid this we use an upload URL to increase the limit to 5 GB.

Reviewer 2: If we can update the request for uploading the blend file into a redirect then we might be able to remove this note. It's worth noting that the size limit for uploading directly to AWS lambda is 6MB

async def get_render(ctx: HttpContext):
  file_name = ctx.req.params['file']

  download_url = await readable_writeable_blend_bucket.file(file_name).download_url(3600)

  ctx.res.headers["Location"] = download_url
  ctx.res.status = 303

  return ctx

@main_api.post("/:blend")
async def write_render(ctx: HttpContext):
  blend_scene_key = ctx.req.params["blend"]

  # Use upload url to get around write limit
  upload_url = await readable_writeable_blend_bucket.file(f"{blend_scene_key}.blend").upload_url()

  resp = requests.post(upload_url, data=ctx.req.data)
  if resp.status_code >= 200 and resp.status_code < 300:
    ctx.res.status = resp.status_code
    ctx.res.body = resp.text
    return ctx

  raw_metadata = {
    "file_format": str(ctx.req.query.get('file_format', ['PNG'])[0]),
    "fps": int(ctx.req.query.get('fps', [0])[0]),
    "device": str(ctx.req.query.get('device', ['GPU'])[0]),
    "engine": str(ctx.req.query.get('engine', ['CYCLES'])[0]),
    "animate": bool(ctx.req.query.get('animate', [False])[0]),
  }

  metadata = bytes(json.dumps(raw_metadata), encoding="utf-8")

  await readable_writeable_blend_bucket.file(f"{blend_scene_key}.metadata.json").write(metadata)

>   return ctx

Reviewer 2: Could this be a put redirect much like how your handled the GET request?

    ctx.res.status = resp.status_code
    ctx.res.body = resp.text
    return ctx

  raw_metadata = {
    "file_format": str(ctx.req.query.get('file_format', ['PNG'])[0]),
    "fps": int(ctx.req.query.get('fps', [0])[0]),
    "device": str(ctx.req.query.get('device', ['GPU'])[0]),
    "engine": str(ctx.req.query.get('engine', ['CYCLES'])[0]),
    "animate": bool(ctx.req.query.get('animate', [False])[0]),
  }

  metadata = bytes(json.dumps(raw_metadata), encoding="utf-8")

  await readable_writeable_blend_bucket.file(f"{blend_scene_key}.metadata.json").write(metadata)

  return ctx

@blend_bucket.on("write", "*")
async def on_written_image(ctx: BucketNotificationContext):
  key_without_extension = ctx.req.key.split(".")[0]

  # Ensure {scene}.metadata.json exists and {scene}.blend
  blend_exists = await readable_writeable_blend_bucket.exists(f"{key_without_extension}.blend")
  metadata_exists = await readable_writeable_blend_bucket.exists(f"{key_without_extension}.metadata.json")

  if not blend_exists or not metadata_exists:
    ctx.res.success = False
    return ctx

  await submittable_renderer_job.submit(
    {
      "key": key_without_extension,
    }
  )

>   return ctx

Reviewer 2: Is the exists check to throw away a second request from this triggering twice?

could we use prefix filtering to ensure this is only triggered for the blend files and assuming the existence of the metadata file (or use the metadata defaults in their absence).
